adapters/Unity/UnityAdapter/GADMAdapterUnity.m

```diff
@@ -123,68 +131,66 @@ - (NSString *)getPlacementID {
 
 #pragma mark - Unity Delegate Methods
 
-- (void)unityAdsPlacementStateChanged:(NSString *)placementId
+- (void)unityAdsPlacementStateChanged:(NSString *)placementID
                              oldState:(UnityAdsPlacementState)oldState
                              newState:(UnityAdsPlacementState)newState {
-  // This callback is not forwarded to the adapter by the GADMAdapterUnitySingleton and the adapter
-  // should use the unityAdsReady: and unityAdsDidError: callbacks to forward Unity Ads SDK state to
-  // Google Mobile Ads SDK.
+  id<GADMAdNetworkConnector> strongNetworkConnector = _networkConnector;
+  if (strongNetworkConnector && [placementID isEqualToString:_placementID]) {
+    if (newState == kUnityAdsPlacementStateNoFill){
+      NSError *errorWithDescription = GADUnityErrorWithDescription(@"NO_FILL");
+      [strongNetworkConnector adapter:self didFailAd:errorWithDescription];
+    } else if (newState == kUnityAdsPlacementStateDisabled) {
+      NSError *errorWithDescription = GADUnityErrorWithDescription(@"PlACEMENT_DISABLED");
+      [strongNetworkConnector adapter:self didFailAd:errorWithDescription];
+    } else if (newState == kUnityAdsPlacementStateNotAvailable) {
+      NSError *errorWithDescription = GADUnityErrorWithDescription(@"PlACEMENT_NOTAVAILABLE");
+      [strongNetworkConnector adapter:self didFailAd:errorWithDescription];
+    }
+  }
 }
 
 - (void)unityAdsDidFinish:(NSString *)placementID withFinishState:(UnityAdsFinishState)state {
   id<GADMAdNetworkConnector> strongNetworkConnector = _networkConnector;
-  if (strongNetworkConnector) {
-    [strongNetworkConnector adapterWillDismissInterstitial:self];
-    [strongNetworkConnector adapterDidDismissInterstitial:self];
+  if (strongNetworkConnector && [placementID isEqualToString:_placementID]) {
+    if (state == kUnityAdsFinishStateCompleted) {
+      [strongNetworkConnector adapterDidDismissInterstitial:self];
+    } else if (state == kUnityAdsFinishStateError) {
+      [strongNetworkConnector adapterWillPresentInterstitial:self];
+      [strongNetworkConnector adapterDidDismissInterstitial:self];
+    } else if (state == kUnityAdsFinishStateSkipped) {
+      [strongNetworkConnector adapterDidDismissInterstitial:self];
+    }
+    [UnityAds removeDelegate:self];
   }
 }
 
 - (void)unityAdsReady:(NSString *)placementID {
   id<GADMAdNetworkConnector> strongNetworkConnector = _networkConnector;
-  if (!_isLoading) {
-    return;
-  }
-
-  if (strongNetworkConnector) {
+  if (strongNetworkConnector && [placementID isEqualToString:_placementID]) {
```

Comment:
What if the initialization call gets us into a READY state before this class gets the init complete callback and register this class as a delegate. Will UnityAds SDK still send the unityAdsReady callback to this object? It seems we risk a race condition where we could miss a READY call.

It would be helpful to understand possible state diagram with placementStateChanged. i.e. will we always go into kUnityAdsPlacementStateWaiting a run loop prior to ready?

Reply:
Not sure what I meant by this comment,  I think I made this before code changed. Consider this resolved.